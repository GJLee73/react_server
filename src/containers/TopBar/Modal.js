/**
 * 로그인 컨테이너
 * @author G1
 * @logs // 18.2.25
 */

import React from 'react';
import {connect} from 'react-redux';
import  {RegisterComponent, LoginComponent } from '../../components/auth';
import {loginRequest,findId,findPwd} from '../../modules/authentication'
import {idRequest,emailRequest,registerRequest, phoneRequest} from '../../modules/register';//todo

import { Find } from './Find';
//import {browserHistory} from 'react-router';

export class Modal extends React.Component{
	constructor(props){
        super(props);
        this.state={
            //register state
             type:'login'}
        this.handleLogin = this.handleLogin.bind(this);
        this.handleRegister = this.handleRegister.bind(this);
        this.changeType=this.changeType.bind(this)
        this.selectType=this.selectType.bind(this)
        this.handleToggle=this.handleToggle.bind(this)
	}
/**
 * 로그인을 한다.
 * @param {string} id 
 * @param {string} pw 
 * @description
 * 	만약 로그인시 이메일 인증되지 않았으면 verify로 리다이렉트 한다.
 *  인증되었으면 home으로
 *  인증이 안됬으면 다시 한다.
 */
	handleLogin(id, pw){
		return this.props.loginRequest(id, pw).then(
			()=>{
				console.log(1)
				if(this.props.loginStatus === "SUCCESS"){

					if(!this.props.verified){
                        console.log('hi')
						this.props.toggle()
						this.props.history.push('/verify')
                  
                    }else{
                        console.log('??')
                        this.props.toggle()
                    }
				} else{
					console.log(2)
				}
			}
		).catch(()=>{
			alert('incorrect userid or password')})
	}
    handleRegister(msg){
		return this.props.registerRequest(msg).then(
			
			() => {
				console.log(this.props.registerStatus)
				if(this.props.registerStatus === "SUCCESS"){
					this.changeType('login')
					return true;
				}
			}
		).catch((err)=>{
			alert('something worng'+err)

			return 0
		
		}
		);
	}
    /**
     * 타입에 따라 컴포넌트를 보여주는 함수.
     */
    selectType(){
        switch(this.state.type){
            case 'login':
            return 	<LoginComponent	onLogin={this.handleLogin} modal={this.props.modal} toggle={this.handleToggle} changeType={this.changeType}/>;
            case 'register':
            return < RegisterComponent onRegister={this.handleRegister} checkId={this.props.idRequest} checkEmail={this.props.emailRequest} checkPhone={this.props.phoneRequest}
                             idCheck={this.props.idCheck}
                              emailCheck={this.props.emailCheck} phoneCheck={this.props.phoneCheck} modal={this.props.modal} toggle={this.handleToggle} changeType={this.changeType}/>
            case 'find':
            return  <Find gottenId={this.props.gottenId} message={this.props.message} findId={this.props.findId} findPwd={this.props.findPwd} changeType={this.changeType}
                        modal={this.props.modal} toggle={this.handleToggle}/>


        }
    }
    /**
     * 타입을 바꾸는 함수.
     * @param {string} type -login, regist, find
     */
    changeType(type){
        console.log(type)
        this.setState({type})
    }
    /**
     * 모달 키고 끄는 함수. 기본적으로 login이 될 수 있게 만든다.
     */
    handleToggle(){
        if(this.state.type !=='login') this.setState({type:'login'})
        this.props.toggle()
    }

	render(){
		return (
			<div>
                {this.selectType()}
                
			</div>
		);
    }	
}


const mapStateToProps = (state) =>{
	return{
		loginStatus: state.authentication.getIn(['login','status']),
        verified: state.authentication.getIn(['status','verified']),
        registerStatus: state.register.regist.status,
		errorCode: state.register.regist.error,
		idCheck:state.register.idCheck,
		emailCheck:state.register.emailCheck,
        phoneCheck:state.register.phoneCheck,
        //find
        gottenId: state.authentication.getIn(['findId','gottenId']),
        message: state.authentication.getIn(['findPwd','message'])
	};
};

const mapDispatchtoProps = (dispatch) => {
	return {
		loginRequest: (id, pw)=>{
			return dispatch(loginRequest(id,pw));
        },
        
        //register

        registerRequest: (id, pw) => dispatch(registerRequest(id, pw)),
		idRequest: (id)=>{
			return dispatch(idRequest(id))
		},
		emailRequest: (email)=>{
			return dispatch(emailRequest(email))
		},
		phoneRequest: (phone)=>{
			return dispatch(phoneRequest(phone))
		},
        findId:(name,email)=>{
            return dispatch(findId(name,email))},
        findPwd:(id,email)=> dispatch(findPwd(id,email))

	};
}

export default connect(mapStateToProps, mapDispatchtoProps)(Modal);