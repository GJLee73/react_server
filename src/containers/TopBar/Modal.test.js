import React from 'react';
import {shallow} from 'enzyme'
import {Modal} from './Modal'

describe('Modal test',()=>{
    var changeType = jest.spyOn(Modal.prototype,'changeType')
    var loginRequest = jest.fn(()=>Promise.resolve())
    var registerRequest=jest.fn(()=>Promise.resolve())
    var toggle = jest.fn()
    var history = {push:jest.fn()}
    global.alert = jest.fn()
    var wrapper= shallow(<Modal loginRequest={loginRequest} toggle={toggle}/>)
    describe('function test',()=>{
        describe('handleLogin test',()=>{
            it('로그인에 성공하면 모달이 꺼진다. ',async()=>{
                wrapper.setProps({loginStatus:"SUCCESS",history,verified:true})
               await wrapper.instance().handleLogin()
                expect(toggle).toHaveBeenCalledTimes(1)
            })
            it('아이디는 존재하나 verified==false일때 /verify로 리다이렉트 된다.',async()=>{
                wrapper.setProps({verified:false})
               await wrapper.instance().handleLogin()
                expect(history.push.mock.calls[0][0]).toBe('/verify')
            })
            it('로그인 실패시 경고창을 띄운다..',async()=>{
                wrapper.setProps({loginRequest:()=>Promise.reject(),loginStatus:'FAILURE'})
                await wrapper.instance().handleLogin()
                expect(global.alert.mock.calls[0][0]).toBe('incorrect userid or password')
            })
    
        })
        describe('handelRegister test ',()=>{
            it('status===Success',async ()=>{
                wrapper.setProps({registerRequest,registerStatus:'SUCCESS'})
                await wrapper.instance().handleRegister()
                expect(registerRequest).toHaveBeenCalledTimes(1)
                expect(changeType).toHaveBeenCalledTimes(1)
            })
            it('status!=success',async ()=>{
                wrapper.setProps({registerStatus:'fail'})
                await wrapper.instance().handleRegister()
                expect(registerRequest).toHaveBeenCalledTimes(2)
                expect(history.push).toHaveBeenCalledTimes(1)
           
            })
            it('registerRequest throw error',async()=>{
                wrapper.setProps({registerRequest:()=>Promise.reject()})
                await wrapper.instance().handleRegister()
                expect(global.alert).toHaveBeenCalledTimes(2)
            })
        })
        describe('selectType test with snapshot',()=>{
            it('type == login',()=>{
                wrapper.setState({type:'login'})
                expect(wrapper).toMatchSnapshot()
            })
            it('type == register',()=>{
                wrapper.setState({type:'register'})
                expect(wrapper).toMatchSnapshot()
            })
            it('type == find',()=>{
                wrapper.setState({type:'find'})
                expect(wrapper).toMatchSnapshot()
            })
        })
        it('changeType test',()=>{
            wrapper.instance().changeType('abc')
            expect(wrapper.state('type')).toBe('abc')
        })
        describe('handleToggle test',()=>{
            it('type !==login',()=>{
                wrapper.instance().handleToggle()
                expect(wrapper.state('type')).toBe('login')
            })
            it('type == login',()=>{
                wrapper.instance().handleToggle()
                expect(toggle).toHaveBeenCalledTimes(4)
            })
           
        })
    })
})