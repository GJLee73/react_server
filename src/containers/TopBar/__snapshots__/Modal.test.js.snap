// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal test function test selectType test with snapshot type == find 1`] = `
<div>
  <Find
    changeType={[Function]}
    toggle={[Function]}
  />
</div>
`;

exports[`Modal test function test selectType test with snapshot type == login 1`] = `
<div>
  <LoginComponent
    changeType={[Function]}
    onLogin={[Function]}
    toggle={[Function]}
  />
</div>
`;

exports[`Modal test function test selectType test with snapshot type == register 1`] = `
<div>
  <RegisterComponent
    changeType={[Function]}
    onRegister={[Function]}
    toggle={[Function]}
  />
</div>
`;
